File: algorithm-c/sorting/mergesort/mergesort.c
#include <stdio.h>

int main()
{
    void pnt(int arr[], int len);
    void merge_array(int* arr1, int len1, int* arr2, int len2);
    void merge_sort(int* arr, int length);

    int array[] = {6, 2, 0, 8, 3, 1, 7, 9, 4, 5};
    pnt(array, 10);
    merge_sort(array, 10);
    pnt(array, 10);
}

void pnt(int arr[], int len)
{
    int i = 0;
    while(i < len)
    {
        printf("%d%s", arr[i++], " ");
    }
    printf("\n");
}

void merge_array(int* arr1, int len1, int* arr2, int len2)
{
    int i, j, k;
    i = j = k = 0;
    int merged_arr[len1 + len2];
    while(i < len1 && j < len2)
    {
        merged_arr[k++] = *(arr1 + i) < *(arr2 + j) ? *(arr1 + i++) : *(arr2 + j++);
    }

    while(i < len1)
    {
        merged_arr[k++] = *(arr1 + i++);
    }

    while(j < len2)
    {
        merged_arr[k++] = *(arr2 + j++);
    }

    int ii = 0;
    while(ii < k)
    {
        *(arr1 + ii) = merged_arr[ii];
        ii++;
    }
}

void merge_sort(int* arr, int length)
{
    if(length > 1)
    {
        int mid = (int)length/2;
        int* arr1 = arr;
        int* arr2 = arr + mid;
        merge_sort(arr1, mid);
        merge_sort(arr2, length - mid);

        merge_array(arr1, mid, arr2, length - mid);
    }
    
}

File: algorithm-js/sorting/merge-sort.js
var arr = [9, 4, 2, 8, 5, 1, 3, 6, 7, 0];

console.log(arr);
console.log(mergeSort(arr));

function mergeSort(array) {
    if(array.length > 1) {
        var midNum = array.length/2;
        var arr1 = array.slice(0, midNum);
        var arr2 = array.slice(midNum);
        arr1 = mergeSort(arr1);
        arr2 = mergeSort(arr2);

        return mergeArray(arr1, arr2);
    } else {
        return array;
    }
}

function mergeArray(arr1, arr2) {
    var i = 0;
    var j = 0;
    var mergedArr = [];
    // while(true) {
    //     if(arr1[i] < arr2[j]) {
    //         mergedArr.push(arr1[i]);
    //         ++i;
    //     } else {
    //         mergedArr.push(arr2[j]);
    //         ++j;
    //     }

    //     if(i == arr1.length) {
    //         mergedArr = mergedArr.concat(arr2.slice(j));
    //         break;
    //     } else if(j == arr2.length) {
    //         mergedArr = mergedArr.concat(arr1.slice(i));
    //         break;
    //     }
    // }

    //improve above code
    while(i < arr1.length && j < arr2.length) {
        arr1[i] < arr2[j] ? mergedArr.push(arr1[i++]) : mergedArr.push(arr2[j++]);
    }
    mergedArr = i == arr1.length ? mergedArr.concat(arr2.slice(j)) : mergedArr.concat(arr1.slice(i));
    return mergedArr;
}
